### ISSUE TITLE
DurableExecutor is ignoring its capacity

### ISSUE THREAD

**Issue by felipeek** [hazelcast/hazelcast#18396]
**Describe the bug**
If you configure a `DurableExecutor` with a capacity of `N` and then you submit `N+1` tasks, you don't get blocked. Also, it seems that you don't need to dispose the futures, ever.

**Expected behavior**
AFAIK you should be blocked, just like as the ScheduledExecutor behaves.

**To Reproduce**

```
final DurableExecutorConfig config = config.getDurableExecutorConfig(EXECUTOR_NAME);
config.setPoolSize(2);
config.setDurability(2);
config.setCapacity(3);
final DurableExecutorService svc = client.getDurableExecutorService(EXECUTOR_NAME);
for (int i = 0; i < 20; ++i) {
  svc.schedule(new MyCallable(), 0, TimeUnit.SECONDS);
}
```

The `for` loop submit 20 parallel tasks that sleep for 30sec. I don't get blocked, but the capacity is 3.

Now if I do the same with a ScheduledExecutor:

```
final IScheduledExecutorService sc = client.getScheduledExecutorService(EXECUTOR_NAME);
final ScheduledExecutorConfig scConfig = config.getScheduledExecutorConfig(EXECUTOR_NAME);
scConfig.setPoolSize(2);
scConfig.setDurability(2);
scConfig.setCapacity(3);

for (int i = 0; i < 20; ++i) {
  sc.submit(new MyCallable());
}
```

I immediately receive a RejectedExecution exception, which is expected.

[this is a single node scenario]

So, basically I noticed that:

- Regardless of how many tasks are currently executing, the DurableExecutor always let me schedule new tasks, even if the capacity is less than the number of tasks that are running.
- Also, for the DurableExecutor, I don't need to ever dispose the futures. This is also an important point: it is not clear to me if the DurableExecutor futures need to be disposed. In contrast to the ScheduleExecutor, the DurableExecutorFuture does not have a `dispose()` method. However, the DurableExecutorService has a `disposeResult()` method. So is it necessary to dispose them?

Thanks

Hazelcast Version: 4.1.1

**Comment by ahmetmircik** [comment]
Hi @felipeek, actually this is not a bug, it is working as expected. The capacity given with `DurableExecutorConfig` is a per-partition capacity(whilst capacity of `ScheduledExecutorConfig` is per-node). By default there are 271 partitions in Hazelcast. Your task is sent to a random partition to execute. If you implement `PartitionAware` interface for your task, you can send all your task into the same partition and see `RejectedExecution` after reaching capacity of 3.

Ragarding dispose you can either use `dispose` or `retrieveAndDisposeResult` otherwise memory leaks.

**Comment by felipeek** [comment]
@ahmetmircik thanks, this makes sense
